### PR TITLE
Session non transferable

### PR DESCRIPTION
**Pull request by equivalent**
locks session to curent ip, browser agent, and session id

therefore if someone obtains session id (e.g. steals cookies)  session will be destroyed => user needs to log back in

you can try this if you export your cookies (using for example [Firefox cookie export plugin](https://addons.mozilla.org/en-us/firefox/addon/export-cookies/) ) and then using `wget --load-cookies=/tmp/cookies.txt http://0.0.0.0:3000/users`

### CODE REVIEW

File: lib/devise_security_extension/hooks/session_non_transferable.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# This hook requires session_limitable 
+
+# After each sign in, native Devise updates current_sign_in_ip and this hook
+# updates unique_session_client.
+# This is only triggered when the user is explicitly set (with set_user)
+# and on authentication. Retrieving the user from session (:fetch) does
+# not trigger it.
+Warden::Manager.after_set_user :except => :fetch do |record, warden, options|
+  if record.respond_to?(:update_current_user_agent!) && warden.authenticated?(options[:scope])
```
Comment (seanknox) [inline lib/devise_security_extension/hooks/session_non_transferable.rb]: `respond_to?`, `kind_of?`, and related methods are generally anti-pattern in Ruby, because you can use duck typing and polymorphism to determine how an object should respond.

**Comment by seanknox** [comment]
Hey @equivalent, there aren't any tests for this new feature. Were you planning on writing some?

File: lib/devise_security_extension/hooks/session_non_transferable.rb (hunk 1 above)
Comment (equivalent) [inline lib/devise_security_extension/hooks/session_non_transferable.rb]: to be honest I'm just fallowing how the gem was originally written, I don't mind to rewrite it so that it is by best practices but I don't want to introduce inconsistency 

my opinion is that It's better to have consistent questionable code rather than inconsistent multiple-ways-to-do-stuff kind of code :)

 please have a look at `/lib/devise_security_extension/hooks/session_limitable.rb` I'm basically just fallowing how this extension was written

**Comment by equivalent** [comment]
hi @seanknox , the thing is the entire gem is lacking  integration tests and there isn't much model testing I can do here :) 

 from what I understand looking at gem issues (like https://github.com/phatworx/devise_security_extension/issues/79 ) the gem promotes writing integration tests in your app. 

Which I kinda see the point as it already happened to me when I upgraded rails version or devise version, where tests were passing in the gem but I ended up with broken feature in my app

But back to your point => 
Basically I need to write some request tests to prove that when different "browser agents" visit the site. But same level of tests are lacking from feature `session_limitable` therefore I need to write those as well 

Anyway I saw your PR on that you are planing to replace test lib to RSpec, That's awesome idea and I'm more than happy to write some request tests for both of these modules after that .

**Comment by traxanos** [comment]
@equivalent 

i will remove "Session non transferable" feature again, because it is the same feature "session_limitable". can you integrate the limitation for user agent and current ip to "session_limitable" so we have only one module

**Comment by traxanos** [comment]
i think the session_limitable feature should be able to configure like this:

config.session_limitable_on_unique_id = BOOL # default true
config.session_limitable_on_user_agent = BOOL
config.session_limitable_on_ip = BOOL

**Comment by equivalent** [comment]
make sence,  I'll do that in next few days :)

**Comment by traxanos** [comment]
or better

config.session_limit_on_*

what do you think, you are the native speaker :)

**Comment by equivalent** [comment]
dude I'm from Slovakia :D I'm glad they let me in UK with my poor English :D 

I like more the 

```
config.session_limitable_on_unique_id = BOOL # default true
config.session_limitable_on_user_agent = BOOL
config.session_limitable_on_ip = BOOL
```

as it more consistent with module name and stuff :)

**Comment by traxanos** [comment]
ok :D
